File: idb_companion/Utility/FBIDBLogger.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBIDBLogger.h"

static NSMutableArray<id<FBControlCoreLogger>> *GlobalLoggers(void)
{
  static dispatch_once_t onceToken;
  static NSMutableArray<id<FBControlCoreLogger>> *loggers;
  dispatch_once(&onceToken, ^{
    loggers = NSMutableArray.array;
  });
  return loggers;
}

static void AddGlobalLogger(id<FBControlCoreLogger> logger)
{
  NSMutableArray<id<FBControlCoreLogger>> *loggers = GlobalLoggers();
  @synchronized (loggers) {
    [loggers addObject:logger];
  }
}

static void RemoveGlobalLogger(id<FBControlCoreLogger> logger)
{
  NSMutableArray<id<FBControlCoreLogger>> *loggers = GlobalLoggers();
  @synchronized (loggers) {
    [loggers removeObject:logger];
  }
}

@interface FBIDBLogger_Operation : NSObject <FBLogOperation>

@property (nonatomic, strong, readonly) id<FBDataConsumer> consumer;
@property (nonatomic, strong, readonly) id<FBControlCoreLogger> logger;
@property (nonatomic, strong, readonly) dispatch_queue_t queue;

@end

@implementation FBIDBLogger_Operation

- (instancetype)initWithConsumer:(id<FBDataConsumer>)consumer logger:(id<FBControlCoreLogger>)logger queue:(dispatch_queue_t)queue
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _consumer = consumer;
  _logger = logger;
  _queue = queue;


  return self;
}

- (FBFuture<NSNull *> *)completed
{
  return [FBMutableFuture.future onQueue:self.queue respondToCancellation:^{
    RemoveGlobalLogger(self.logger);
    return FBFuture.empty;
  }];
}

- (NSString *)futureType
{
  return @"companion_log";
}

@end

@implementation FBIDBLogger

#pragma mark Initializers

+ (dispatch_queue_t)loggerQueue
{
  static dispatch_once_t onceToken;
  static dispatch_queue_t queue;
  dispatch_once(&onceToken, ^{
    queue = dispatch_queue_create("com.facebook.idb.logger", DISPATCH_QUEUE_SERIAL);
  });
  return queue;
}

+ (instancetype)loggerWithUserDefaults:(NSUserDefaults *)userDefaults
{
  BOOL debugLogging = [[userDefaults stringForKey:@"-log-level"].lowercaseString isEqualToString:@"info"] ? NO : YES;
  id<FBControlCoreLogger> systemLogger = [FBControlCoreLogger systemLoggerWritingToStderr:YES withDebugLogging:debugLogging];
  NSMutableArray<id<FBControlCoreLogger>> *loggers = [NSMutableArray arrayWithObject:[FBControlCoreLogger systemLoggerWritingToStderr:YES withDebugLogging:debugLogging]];

  NSError *error;
  NSString *logFilePath = [userDefaults stringForKey:@"-log-file-path"];
  if (logFilePath) {
    NSURL *logFileURL = [NSURL fileURLWithPath:logFilePath];
    if (![NSFileManager.defaultManager createDirectoryAtURL:logFileURL.URLByDeletingLastPathComponent withIntermediateDirectories:YES attributes:@{} error:&error]) {
      [systemLogger.error logFormat:@"Couldn't create log directory at %@: %@", logFileURL.URLByDeletingLastPathComponent, error];
      exit(1);
    }

    int fileDescriptor = open(logFileURL.path.UTF8String, O_WRONLY | O_APPEND | O_CREAT);
    if (!fileDescriptor) {
      [systemLogger.error logFormat:@"Couldn't create log file at %@ %s", logFileURL.path, strerror(errno)];
      exit(1);
    }

    [loggers addObject:[FBControlCoreLogger loggerToFileDescriptor:fileDescriptor closeOnEndOfFile:YES]];
  }
  FBIDBLogger *logger = [[[FBIDBLogger alloc] initWithLoggers:loggers] withDateFormatEnabled:YES];
  FBControlCoreGlobalConfiguration.defaultLogger = logger;

  return logger;
}

#pragma mark FBCompositeLogger

- (instancetype)initWithLoggers:(NSArray<id<FBControlCoreLogger>> *)loggers
{
  self = [super initWithLoggers:loggers];
  if (!self) {
    return nil;
  }

  return self;
}

- (NSArray<id<FBControlCoreLogger>> *)loggers
{
  NSMutableArray<id<FBControlCoreLogger>> *global = GlobalLoggers();
  NSMutableArray<id<FBControlCoreLogger>> *all = [[super loggers] mutableCopy];
  @synchronized (global)
  {
    [all addObjectsFromArray:global];
  }
  return all;
}

#pragma mark Public Methods

- (FBFuture<id<FBLogOperation>> *)tailToConsumer:(id<FBDataConsumer>)consumer
{
  dispatch_queue_t queue = FBIDBLogger.loggerQueue;
  return [FBFuture onQueue:queue resolveValue:^(NSError **_) {
    id<FBControlCoreLogger> logger = [FBControlCoreLogger loggerToConsumer:consumer];
    id<FBLogOperation> operation =  [[FBIDBLogger_Operation alloc] initWithConsumer:consumer logger:logger queue:queue];
    AddGlobalLogger(logger);
    return operation;
  }];
}

@end

File: CompanionLib/Utility/FBIDBLogger.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Darwin
@preconcurrency import FBControlCore
import Foundation

nonisolated(unsafe) private let globalLoggers: NSMutableArray = NSMutableArray()
private let globalLoggersLock = NSLock()

private func addGlobalLogger(_ logger: FBControlCoreLogger) {
  globalLoggersLock.lock()
  globalLoggers.add(logger)
  globalLoggersLock.unlock()
}

private func removeGlobalLogger(_ logger: FBControlCoreLogger) {
  globalLoggersLock.lock()
  globalLoggers.remove(logger)
  globalLoggersLock.unlock()
}

private class FBIDBLoggerOperation: NSObject, FBLogOperation {
  let consumer: FBDataConsumer
  let logger: FBControlCoreLogger
  let queue: DispatchQueue

  init(consumer: FBDataConsumer, logger: FBControlCoreLogger, queue: DispatchQueue) {
    self.consumer = consumer
    self.logger = logger
    self.queue = queue
    super.init()
  }

  var completed: FBFuture<NSNull> {
    let logger = self.logger
    let cls = unsafeBitCast(NSClassFromString("FBMutableFuture")!, to: NSObject.Type.self)
    let mutableFuture = cls.perform(NSSelectorFromString("future"))!.takeUnretainedValue() as! FBFuture<NSNull>
    return mutableFuture.onQueue(
      self.queue,
      respondToCancellation: {
        removeGlobalLogger(logger)
        return FBFuture<NSNull>.empty()
      })
  }

  var operationType: String {
    "companion_log"
  }
}

@objc public final class FBIDBLogger: FBCompositeLogger {

  private static let loggerQueue: DispatchQueue = DispatchQueue(label: "com.facebook.idb.logger")

  @objc public static func logger(withUserDefaults userDefaults: UserDefaults) -> FBIDBLogger {
    let debugLogging = userDefaults.string(forKey: "-log-level")?.lowercased() == "info" ? false : true
    let systemLogger = FBControlCoreLoggerFactory.systemLoggerWriting(toStderr: true, withDebugLogging: debugLogging)
    let loggers: NSMutableArray = NSMutableArray(object: systemLogger)

    let logFilePath = userDefaults.string(forKey: "-log-file-path")
    if let logFilePath {
      let logFileURL = URL(fileURLWithPath: logFilePath)
      do {
        try FileManager.default.createDirectory(at: logFileURL.deletingLastPathComponent(), withIntermediateDirectories: true, attributes: [:])
      } catch {
        systemLogger.error().log("Couldn't create log directory at \(logFileURL.deletingLastPathComponent()): \(error)")
        exit(1)
      }

      let fileDescriptor = open(logFileURL.path, O_WRONLY | O_APPEND | O_CREAT)
      if fileDescriptor == 0 {
        systemLogger.error().log("Couldn't create log file at \(logFileURL.path) \(String(cString: strerror(errno)))")
        exit(1)
      }

      loggers.add(FBControlCoreLoggerFactory.logger(toFileDescriptor: fileDescriptor, closeOnEndOfFile: true))
    }
    let logger = FBIDBLogger(loggers: loggers as! [FBControlCoreLogger]).withDateFormatEnabled(true) as! FBIDBLogger
    FBControlCoreGlobalConfiguration.defaultLogger = logger

    return logger
  }

  @objc public override init(loggers: [FBControlCoreLogger]) {
    super.init(loggers: loggers)
  }

  @objc public override var loggers: [FBControlCoreLogger] {
    var all = super.loggers
    globalLoggersLock.lock()
    let global = globalLoggers as! [FBControlCoreLogger]
    globalLoggersLock.unlock()
    all.append(contentsOf: global)
    return all
  }

  @objc public func tailToConsumer(_ consumer: FBDataConsumer) -> FBFuture<AnyObject> {
    fbFutureFromAsync { [self] in
      try await tailToConsumerAsync(consumer) as AnyObject
    }
  }

  public func tailToConsumerAsync(_ consumer: FBDataConsumer) async throws -> FBLogOperation {
    let queue = FBIDBLogger.loggerQueue
    return await withCheckedContinuation { (continuation: CheckedContinuation<FBLogOperation, Never>) in
      queue.async {
        let logger = FBControlCoreLoggerFactory.logger(to: consumer)
        let operation = FBIDBLoggerOperation(consumer: consumer, logger: logger, queue: queue)
        addGlobalLogger(logger)
        continuation.resume(returning: operation)
      }
    }
  }
}
